core/libp2p/transport/tcp/tcp_connection.cpp

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/transport/tcp/tcp_connection.hpp"
+
+#include <sstream>
+
+namespace libp2p::transport {
+
+  TcpConnection::TcpConnection(boost::asio::io_context &context)
+      : socket_(context) {}
+
+  TcpConnection::TcpConnection(boost::asio::ip::tcp::socket socket)
+      : socket_(std::move(socket)) {}
+
+  outcome::result<std::vector<multi::Multiaddress>>
+  TcpConnection::getObservedAddresses() const {
+    try {
+      auto &&remote = socket_.remote_endpoint();
+      auto &&addr = remote.address().to_string();
+      auto &&port = remote.port();
+      std::ostringstream s{};
+      s << "/ip4/" << addr << "/tcp/" << port;
+      OUTCOME_TRY(ma, multi::Multiaddress::create(s.str()));
+      return std::vector<multi::Multiaddress>{ma};
+    } catch (const boost::system::system_error &e) {
+      return e.code();
+    }
+  }
```

Comment:
You can use `try` instead of the function's curly braces. Looks prettier, I think

Reply:
Didn't get. Example?

Reply:
```
int foo() try {
  throw exception();
} catch {
  std::terminate();
}
```